commons/src/main/java/org/infinispan/commons/api/functional/FunctionalMap.java

```diff
@@ -0,0 +1,634 @@
+       *    instead of forcing users to iterate over the keys to lookup and call
+       *    get individually since Infinispan can do things more efficiently.
+       *    </li>
+       *    <li>Since this is a write-only operation, no entry attributes can be
+       *    queried, hence the only reasonable thing can be returned is Void.
+       *    </li>
+       *    <li>It still makes sense to return a CloseableIterator<Void> instead of
+       *    CompletableFuture<Void> in case the user wants to do something as each
+       *    operation completes, e.g. keep track of progress.
+       *    </li>
+       * </ul>
+       *
+       * @param entries the key/value pairs associated with each of the
+       *             {@link WriteEntryView} passed in the function callbacks
+       * @param f operation that consumes a value associated with a key in the
+       *          entries collection and the {@link WriteEntryView} associated
+       *          with that key in the cache
+       * @return a {@link CloseableIterator} that can be navigated to follow progress
+       *         as operations get invoked
+       */
+      CloseableIterator<Void> evalMany(Map<? extends K, ? extends V> entries, BiConsumer<V, WriteEntryView<V>> f);
+
+      /**
+       * Evaluate a write-only {@link Consumer} operation with the
+       * {@link WriteEntryView} of the value associated with the key, for each
+       * of the keys in the set passed in, and returns a
+       * {@link CloseableIterator} that allows navigating completion
+       * of the operation executions.
+       *
+       * This method can be used to implement operations such as
+       * {@link javax.cache.Cache#removeAll(Set)}.
+       *
+       * DESIGN RATIONALE:
+       * <ul>
+       *    <li>It makes sense to expose global operation like this
+       *    instead of forcing users to iterate over the keys to lookup and call
+       *    get individually since Infinispan can do things more efficiently.
+       *    </li>
+       *    <li>Since this is a write-only operation, no entry attributes can be
+       *    queried, hence the only reasonable thing can be returned is Void.
+       *    </li>
+       *    <li>It still makes sense to return a CloseableIterator<Void> instead of
+       *    CompletableFuture<Void> in case the user wants to do something as each
+       *    operation completes, e.g. keep track of progress.
+       *    </li>
+       * </ul>
+       *
+       * @param keys the keys associated with each of the {@link WriteEntryView}
+       *             passed in the function callbacks
+       * @param f operation that the {@link WriteEntryView} associated with
+       *          one of the keys passed in
+       * @return a {@link CloseableIterator} that can be navigated to follow progress
+       *         as operations get invoked
+       */
+      CloseableIterator<Void> evalMany(Set<? extends K> keys, Consumer<WriteEntryView<V>> f);
```

Comment:
Sorry I think I missed the value matcher.  But I am thinking either way we need a retry consumer argument possibly as well?